File: postm/middlewares/authentication.py
from flask import request
from os import getenv

from jwt import decode
from jwt import ExpiredSignatureError

from postm.services.users import UsersService

class AuthMiddleware(object):

    def __init__(self) -> None:
        self.service = UsersService()

    def auth(self) -> tuple[dict, int] | None:
        authorization = request.headers.get("Authorization")

        if not authorization:
            return {
                "error": "an authentication token is required"
            }, 400
        
        parts = authorization.split(" ")

        if len(parts) != 2 or parts[0] != "Bearer":
            return {
                "error": "authentication token needs to be a jwt"
            }, 400
        
        tokenEncoded = parts[1]

        try:
            payload: dict = decode(
                jwt = tokenEncoded,
                key = getenv("JWT_KEY"),
                algorithms = ["HS256"]
            )

            userId = payload.get("id", "")

            user = self.service.findById(userId)

            if not user:
                return {
                    "error": f"user { userId } not found"
                }, 401
            
            request.user = user
            
        except ExpiredSignatureError as error:
            return {
                "error": "token expired"
            }, 401  

        except Exception as error:
            return {
                "error": str(error)
            }, 500

File: src/middlewares/authentication.ts
import { Request, Response, NextFunction } from "express";
import { z } from "zod";
import jwt from "jsonwebtoken";
import env from "../env";

const headerSchema = z.object({
    authorization: z.string({ message: "Authorization header required" })
});

const payloadSchema = z.object({
    id: z.string({ message: "id required" })
});

export default class Authentication {

    public auth(request: Request, response: Response, next: NextFunction) {

        try {
            const { authorization } = headerSchema.parse(request.headers);
    
            const parts = authorization.split(" ");
    
            if (parts.length < 2 || parts[0] != "Bearer") {
                response.status(403).json({
                    error: "authentication token needs to be a jwt"
                });
    
                return;
            }
            const tokenEncoded = parts[1];
    
            const { id } = payloadSchema.parse(
                jwt.verify(tokenEncoded, env.JWT_KEY)
            );

            (request as any).userId = id;

            next();

        } catch (error: any) {

            if (error instanceof z.ZodError) {
                response.status(403).json({
                    message: "token bad formated"
                });
                return;
            }

            response.status(500).json({
                message: error.message
            });

        }

    }


}